base/tks/src/main/java/org/dogtagpki/server/tks/servlet/TokenServlet.java

```diff
@@ -3071,47 +3167,75 @@ private ArrayList<String> calculateServerSideKeygenValues(String useSoftToken, S
         if (useSoftToken.equals("true")) {
             logger.debug(method + " key encryption key generated on internal");
             desKey = protocol.generateSymKey("internal");
+	    if(protocolLevel == 3) {
+                //still do the des key as a backup later
+                aesKey = protocol.generateAESSymKey("internal",128);
+            }
             //cfu audit here? sym key gen done
         } else {
             logger.debug("TokenServlet: key encryption key generated on " + selectedToken);
             desKey = protocol.generateSymKey(selectedToken);
+            if(protocolLevel == 3) {
+                aesKey = protocol.generateAESSymKey(selectedToken,128);
+            }
         }
-        if (desKey == null) {
-            throw new EBaseException(method + "can't generate key encryption key");
+        if (desKey == null && protocolLevel == 1) {
+            throw new EBaseException(method + "can't generate DES key encryption key");
         }
 
+	if (aesKey == null && protocolLevel == 3) {
+            throw new EBaseException(method + "can't generate AES key encryption key");
+	}
+
         /*
          * ECBencrypt actually takes the 24 byte DES2 key
          * and discard the last 8 bytes before it encrypts.
          * This is done so that the applet can digest it
          */
 
-
-       // protocol.wrapSessionKey(tokenName, sessionKey, wrappingKey)
+        /*
+	 * Now we also must account for AES session keys
+	 *
+	 */
 
         byte[] encDesKey = protocol.ecbEncrypt(kekSessionKey, desKey, selectedToken);
+	byte[] encAesKey = protocol.ecbEncrypt(kekSessionKey, aesKey, selectedToken);
 
         String kek_wrapped_desKeyString =
                 org.mozilla.jss.netscape.security.util.Utils.SpecialEncode(encDesKey);
 
-        //logger.debug(method + "kek_wrapped_desKeyString: " + kek_wrapped_desKeyString);
+        String kek_wrapped_aesKeyString =
+                org.mozilla.jss.netscape.security.util.Utils.SpecialEncode(encAesKey);
 
         values.add(kek_wrapped_desKeyString);
```

Comment:
might be a good idea to do teh values.add of kek_wrapped_desKeyString and kek_wrapped_aesKeyString at similar location, either here, or later.